compiler/vm/vmlinker.nim

```diff
@@ -0,0 +1,57 @@
+## Implements types and procedures related to mapping symbols to their in-VM
+## IDs.
+
+import
+  std/[
+    strutils,
+    tables
+  ],
+  compiler/ast/[
+    ast_types
+  ]
+
+from compiler/ast/ast import id
+from compiler/ast/idents import cmpIgnoreStyle
+
+type
+  LinkIndex* = uint32
+    ## Identifies a linker-relevant entity. There are three namespaces, one
+    ## for procedures, one for globals, and one for constants -- which
+    ## namespace an index is part of is stored separately.
+
+  IdentPattern* = distinct string
+    ## A matcher pattern for a fully qualified symbol identifier
+
+  LinkerData* = object
+    ## All data required for mapping symbols to the entities they represent
+    ## inside the VM (procedures, globals, constants).
+    symToIndexTbl*: Table[int, LinkIndex]
+      ## the link table. Keeps track of all symbols known to the linker
+    callbackKeys*: seq[IdentPattern]
+      ## the matcher patterns used for implementing overrides
```

Comment:
an invariant we need is that all `callbackKeys` entries, if linked, must have an entry at the corresponding `int` in `symToIndexTbl`? or perhaps, if a key exists in `symToIndexTbl` that overlaps with the index of `callbackKeys`, that must be the `LinkIndex` for that `IdentPattern`?

If that's the case, I think it might be worth noting this invariant, as it indicates not only the cardinality/relationship, but clearly hints as to how data should be organized with the overall type.

Reply:
> an invariant we need is that all `callbackKeys` entries, if linked, must have an entry at the corresponding `int` in `symToIndexTbl`?

Not directly, no. There exists no relation between `callbackKeys` and `LinkIndex` (and thus `symToIndexTbl`).

Once the necessary progress has been made in the code generators, `symToIndexTbl` is going be replaced with three `OrdinalSeq`s (one for each relevant symbol kind), where each uses a key type that corresponds to the ID of a `Store` stored elsewhere. That way, the relationship is going to be more clear.

Reply:
apologies, had a dyslexic moment, I thought I read the the second `lookup` procedure delegating to the first but that's clearly not the case.